File: src/boj/Main16918_봄버맨.py
R, C, N = map(int, input().split())
map = []
time = [[0] * C for _ in range(R)]
dy = [-1, 1, 0, 0]
dx = [0, 0, -1, 1]
for i in range(R):
    map.append(list(input()))
    for j in range(C):
        if map[i][j] == 'O':
            # 폭탄의 남은 시간을 2초로 설정
            time[i][j] = 2


def timepass():
    # 모든 폭탄의 남은 시간을 1초씩 감소
    for r in range(R):
        for c in range(C):
            if time[r][c] > 0:
                time[r][c] -= 1

    for r in range(R):
        for c in range(C):
            if time[r][c] == 0 and map[r][c] == 'O':
                # 폭탄을 터트린다.
                map[r][c] = '.'

                # 4방향에 있는 폭탄을 터트린다.
                for i in range(4):
                    nr = r + dy[i]
                    nc = c + dx[i]
                    if 0 <= nr < R and 0 <= nc < C and time[nr][nc] > 0:
                        time[nr][nc] = 0
                        map[nr][nc] = '.'


# 1초 동안 봄버맨은 아무것도 하지 않는다.
N -= 1

# 3과 4를 반복한다.
while N > 0:
    # 3 폭탄이 설치되어 있지 않은 모든 칸에 폭탄을 설치한다.
    N -= 1
    timepass()
    for i in range(R):
        for j in range(C):
            if map[i][j] == '.':
                map[i][j] = 'O'
                time[i][j] = 3
    if N <= 0:
        break

    # 4 설치된 폭탄이 폭발한다.
    N -= 1
    timepass()

# 결과 출력
for i in range(R):
    print(''.join(map[i]))

File: src/boj/Main16918_봄버맨.java
package boj;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;

public class Main16918_봄버맨 {

    static int R, C, N; // 1 ~ 200
    static int[][] bomb; // 폭탄 상태
    static char[][] map; // 맵의 상태
    static final int[] dx = { 0, 0, -1, 1 }; // 상하좌우 변위
    static final int[] dy = { -1, 1, 0, 0 }; // 상하좌우 변위
    static String[] s;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        s = br.readLine().split(" ");

        R = Integer.parseInt(s[0]);
        C = Integer.parseInt(s[1]);
        N = Integer.parseInt(s[2]);

        map = new char[R][C];
        bomb = new int[R][C];

        for (int i = 0; i < R; i++) {
            map[i] = br.readLine().toCharArray();

            for (int j = 0; j < C; j++) {
                if (map[i][j] == 'O') { // 1. 일부 칸에 폭탄을 설치한다.
                    bomb[i][j] = 4;
                }
            }
        }

        N--; // 2. 봄버맨은 아무것도 하지 않는다.
        time();

        while (N > 0) {
            // 3. 폭탄이 설치되지 않은 칸에 폭탄을 설치한다.
            time();
            N--;
            setBomb();

            if (N == 0)
                break;


            // 4. 폭탄을 폭발시킨다.
            time();
            boom();
            N--;
        }

        print();
    }

    private static void print() {
        for (int i = 0; i < R; i++) {
            for (int j = 0; j < C; j++) {
                System.out.print(map[i][j]);
            }
            System.out.println();
        }
    }

    private static void boom() {
        int nRow, nCol;

        for (int i = 0; i < R; i++) {
            for (int j = 0; j < C; j++) {
                if (bomb[i][j] == 1) {

                    // 해당 폭탄이 터진다.
                    bomb[i][j] = 0;
                    map[i][j] = '.';

                    // 4방향의 폭탄을 제거한다.
                    for (int k = 0; k < 4; k++) {
                        nRow = i + dy[k];
                        nCol = j + dx[k];

                        if (nRow >= 0 && nRow < R && nCol >= 0 && nCol < C && bomb[nRow][nCol] != 1) {
                            bomb[nRow][nCol] = 0;
                            map[nRow][nCol] = '.';
                        }
                    }
                }
            }
        }
    }

    private static void time() {
        for (int i = 0; i < R; i++) {
            for (int j = 0; j < C; j++) {
                if (bomb[i][j] > 0) {
                    bomb[i][j]--;
                }
            }
        }
    }

    private static void setBomb() {
        for (int i = 0; i < R; i++) {
            for (int j = 0; j < C; j++) {
                if (bomb[i][j] == 0) {

                    // 폭탄 시간을 추가한다.
                    bomb[i][j] = 4;

                    // 폭탄을 맵에 추가한다.
                    map[i][j] = 'O';
                }
            }
        }
    }
}



